File: StopWatchAppAndroid/app/src/main/java/com/amay077/stopwatchapp/models/StopWatchModel.java
package com.amay077.stopwatchapp.models;

import com.annimon.stream.Optional;
import com.annimon.stream.Stream;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.subjects.BehaviorSubject;
import io.reactivex.subjects.Subject;


/**
 * ストップウォッチの機能を実装したロジッククラス
 */
public class StopWatchModel {
    // ストップウォッチの状態を更新＆通知するための Subject 群

    private final BehaviorSubject<Long> _time = BehaviorSubject.<Long>createDefault(0L); // タイマー時間
    private final BehaviorSubject<Boolean> _isRunning = BehaviorSubject.<Boolean>createDefault(false); // 実行中か？
    private final BehaviorSubject<List<Long>> _laps = BehaviorSubject.<List<Long>>createDefault(Collections.<Long>emptyList()); // 経過時間群
    private final BehaviorSubject<Boolean> _isVisibleMillis = BehaviorSubject.<Boolean>createDefault(true); // ミリ秒表示するか？

    // スレッドを超えて値を更新(onNext)するための SerializedSubject 群
    private final Subject<Long> _timeSerialized = _time.toSerialized();
    private final Subject<Boolean> _isRunningSerialized = _isRunning.toSerialized();

    // タイマーの購読状況
    private Disposable _timerSubscription = null;

    private String _formattedFastestLap;
    private String _formattedWorstLap;

    // Model として公開するプロパティ
    // Subject をそのまま公開したくないので Observable<> にしている
    public final Observable<String> formattedTime;
    public final Observable<List<String>> formattedLaps;
    public final Observable<Boolean> isRunning = _isRunning;
    public final Observable<Boolean> isVisibleMillis = _isVisibleMillis;

    private final AtomicLong _startTime = new AtomicLong(0L);

    private String formatTime(long t, String format) {
        final SimpleDateFormat sdf = new SimpleDateFormat(format, Locale.getDefault());
        return sdf.format(new Date(t));
    }

    public StopWatchModel() {
        Observable<String> _timeFormat = _isVisibleMillis.map(visible -> visible ? "mm:ss.SSS" : "mm:ss");

        formattedTime = Observable.combineLatest(_time, _timeFormat,
                (t, f) -> formatTime(t, f));

        formattedLaps = Observable.combineLatest(_laps, _timeFormat,
                (laps, f) ->  {
                    Stream<Long> stream = Stream.of(laps);

                    // NOTE ここで max/min を得るのは副作用なのでバッドパターンになり得るが、
                    //      今回は用途が限定的なのでご勘弁。
                    Optional<Long> max = stream.max((x, y)-> x.compareTo(y));
                    _formattedWorstLap = formatTime(max.orElse(0L), f);
                    Optional<Long> min = Stream.of(laps).min((x, y)-> x.compareTo(y));
                    _formattedFastestLap = formatTime(min.orElse(0L), f);

                    return Stream.of(laps).map(l -> formatTime(l, f)).toList();
                });
    }

    /**
     * 計測を開始または終了する(time プロパティの更新を開始する)
     */
    public void startOrStop() {
        if (!_isRunning.getValue()) {
            start();
        } else {
            stop();
        }
    }

    private void start() {
        if (_isRunning.getValue()) {
            stop();
        }

        _timerSubscription =
                Observable.interval(100, TimeUnit.MILLISECONDS)
                        .compose((Observable<Long> x) -> {
                            // 開始時に Laps をクリア、実行中フラグをON
                            _laps.onNext(Collections.<Long>emptyList());
                            _isRunningSerialized.onNext(true);
                            _startTime.set(System.currentTimeMillis());
                            return x;
                        })
                        .subscribe((Long notUse) -> {
                            // タイマー値を通知
                            updateTime();
                        });
    }

    private void updateTime() {
        _timeSerialized.onNext(System.currentTimeMillis() - _startTime.get());
    }

    private void stop() {

        updateTime();

        if (_timerSubscription != null) {
            _timerSubscription.dispose();
            _timerSubscription = null;
        }

        // 実行終了を通知
        _isRunningSerialized.onNext(false);
    }

    /**
     * 経過時間を記録する(経過時間を laps プロパティに追加する)
     */
    public void lap() {
        final List<Long> laps = _laps.getValue();
        final List<Long> newLaps = new ArrayList<>();

        newLaps.addAll(laps);

        long totalLap = 0L;
        for (Long lap:laps) {
            totalLap += lap;
        }

        updateTime();
        newLaps.add(_time.getValue() - totalLap);

        _laps.onNext(Collections.unmodifiableList(newLaps));
    }

    public String getFormattedFastestLap() {
        return _formattedFastestLap;
    }

    public String getFormattedWorstLap() {
        return _formattedWorstLap;
    }

    /**
     * 小数点以下の表示有無を切り替える（isVisibleMillis プロパティを toggle する）
     */
    public void toggleVisibleMillis() {
        _isVisibleMillis.onNext(!_isVisibleMillis.getValue());
    }

    public void unsubscribe() {
        stop();
    }
}

File: StopWatchAppXamarinForms/StopWatchAppXamarinForms/Models/StopWatchModel.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Reactive.Concurrency;
using System.Reactive.Linq;
using System.Threading;
using Reactive.Bindings;

namespace StopWatchAppXamarinForms.Models
{
    /// <summary>
    /// ストップウォッチの機能を実装したロジッククラス
    /// </summary>
    public class StopWatchModel : IStopWatchModel
    {
        // ストップウォッチの状態を更新＆通知するための ReactiveProperty 群(要は Subject)
        private readonly ReactiveProperty<long> _time = new ReactiveProperty<long>(Scheduler.Default, 0L); // タイマー時間

        private readonly ReactiveProperty<bool> _isRunning = new ReactiveProperty<bool>(Scheduler.Default, false); // 実行中か？

        private readonly ReactiveProperty<IList<long>> _laps = new ReactiveProperty<IList<long>>(new List<long>()); // 経過時間群

        private readonly ReactiveProperty<bool> _isVisibleMillis = new ReactiveProperty<bool>(Scheduler.Default, true); // ミリ秒表示するか？
        private readonly ReactiveProperty<string> _timeFormat;

        // Model として公開するプロパティ
        public ReadOnlyReactiveProperty<string> FormattedTime { get; } // フォーマットされたタイマー時間

        public ReadOnlyReactiveProperty<bool> IsRunning { get; } // 実行中か？
        public ReadOnlyReactiveProperty<IEnumerable<string>> FormattedLaps { get; } // フォーマットされた経過時間群
        public ReadOnlyReactiveProperty<bool> IsVisibleMillis { get; } // ミリ秒表示するか？
        public ReadOnlyReactiveProperty<string> FormattedFastestLap { get; }
        public ReadOnlyReactiveProperty<string> FormattedWorstLap { get; }

        // タイマーの購読状況
        private IDisposable _timerSubscription = null;

        private long _startTime;

        public StopWatchModel()
        {
            Debug.WriteLine("StopWatchModel ctor");

            IsRunning = _isRunning.ToReadOnlyReactiveProperty(eventScheduler: Scheduler.Default);
            IsVisibleMillis = _isVisibleMillis.ToReadOnlyReactiveProperty(eventScheduler: Scheduler.Default);

            _timeFormat = IsVisibleMillis
                .Select(v => v ? @"mm\:ss\.fff" : @"mm\:ss")
                .ToReactiveProperty(raiseEventScheduler: Scheduler.Default);

            FormattedTime = _time.CombineLatest(_timeFormat, (time, format) =>
                    TimeSpan.FromMilliseconds(time).ToString(format))
                .ToReadOnlyReactiveProperty(eventScheduler: Scheduler.Default);

            FormattedLaps = _laps.CombineLatest(_timeFormat,
                    (laps, f) => laps.Select((x, i) => TimeSpan.FromMilliseconds(x).ToString(f)))
                .ToReadOnlyReactiveProperty(eventScheduler: Scheduler.Default);

            FormattedFastestLap = _laps.CombineLatest(_timeFormat, (laps, format) =>
                {
                var fastest = laps.Count > 0 ? laps.Min() : 0;
                    return TimeSpan.FromMilliseconds(fastest).ToString(format);
                })
                .ToReadOnlyReactiveProperty(eventScheduler: Scheduler.Default);

            FormattedWorstLap = _laps.CombineLatest(_timeFormat, (laps, format) =>
                {
                    var worst = laps.Count > 0 ? laps.Max() : 0;
                    return TimeSpan.FromMilliseconds(worst).ToString(format);
                })
                .ToReadOnlyReactiveProperty(eventScheduler: Scheduler.Default);
        }

        /// <summary>
        /// 計測を開始または終了する(time プロパティの更新を開始する)
        /// </summary>
        public void StartOrStop()
        {
            if (!_isRunning.Value)
            {
                Start();
            }
            else
            {
                Stop();
            }
        }

        private void Start()
        {
            if (_isRunning.Value)
            {
                Stop();
            }

            // 開始時に Laps をクリア、実行中フラグをON
            // RxJava の compose がないので「購読が開始された時」でないのが微妙…
            _laps.Value = new List<long>();
            ;
            _isRunning.Value = true;
            _startTime = DateTime.Now.Ticks;

            _timerSubscription =
                Observable.Interval(TimeSpan.FromMilliseconds(100))
                          .Subscribe(_ => UpdateTime()); // タイマー値を通知
        }

        private void Stop()
        {
            if (_timerSubscription != null)
            {
                _timerSubscription.Dispose();
                _timerSubscription = null;
            }

            UpdateTime();

            // 実行終了を通知
            _isRunning.Value = false;
        }

        void UpdateTime()
        {
            _time.Value = (DateTime.Now.Ticks - _startTime) / 10000;
        }

        /// <summary>
        /// 経過時間を記録する(経過時間を laps プロパティに追加する)
        /// </summary>
        public void Lap()
        {
            var laps = _laps.Value;
            var newLaps = new List<long>();

            newLaps.AddRange(laps);

            var totalLap = 0L;
            foreach (var lap in laps)
            {
                totalLap += lap;
            }

            UpdateTime();
            newLaps.Add(_time.Value - totalLap);

            _laps.Value = newLaps;
        }

        /// <summary>
        /// 小数点以下の表示有無を切り替える（isVisibleMillis プロパティを toggle する）
        /// </summary>
        public void ToggleVisibleMillis()
        {
            _isVisibleMillis.Value = !_isVisibleMillis.Value;
        }

        #region IDisposable implementation

        public void Dispose()
        {
            Stop();
        }

        #endregion
    }
}